### PR TITLE
Implement server-side connection pooling

### PR DESCRIPTION
**Pull request by 1st1**
ToDo:

- [x] Figure out mypy covariant/contravariant errors
- [ ] Add comments while I still understand this.
- [x] Add a couple more simulations.
- [x] Implement GC-ing of unused connections. This can be done in a separate PR.
- [x] Integrate into `edgecon.pyx`.

--- 

[Latest report](https://gistcdn.githack.com/1st1/17a5823792bb864d87bccaec2a0fb160/raw/76ccbda611c695772627dcd64fa8d031f6538635/test4.html).

### CODE REVIEW

**Review by elprans — COMMENTED** [review]
I've done a superficial pass and left some comments around bits that aren't actual pool.  Will try to look at the pool closer tomorrow when fresh.

File: edb/server/mng_port/edgecon.pyx [hunk 1]
```diff
@@ -1308,103 +1349,93 @@ cdef class EdgeConnection:
...
-                if side_effects & dbview.SideEffects.SystemConfigChanges:
-                    await self.get_backend().pgcon.signal_sysevent(
-                        'system-config-changes')
-
-            self.write(self.make_command_complete_msg(query_unit))
-
-            if process_sync:
-                self.write(self.pgcon_last_sync_status())
-                self.flush()
+                await conn.parse_execute(
```
[the remark this answers is not in the document]
Comment (1st1) [reply edb/server/mng_port/edgecon.pyx:1389]: The diff looks confusing I guess, but no, we'll continue to use them just as we used them before. It's just we will no longer do Parse in Postgres on Parse in EdgeDB; that shifts to EdgeDB's Execute step now.

File: edb/server/bootstrap.py [hunk 2]
```diff
@@ -1204,8 +1176,11 @@ async def _bootstrap(
             f"SET ROLE {qi(args['default_database_user'])};",
         )
 
-    if args['default_database']:
-        await _ensure_edgedb_database(
+    if (
+        args['default_database']
+        and args['default_database'] != edbdef.EDGEDB_SUPERUSER
```
[the remark this answers is not in the document]
Comment (1st1) [reply edb/server/bootstrap.py:1181]: 🙏

File: edb/server/compiler/compiler.py [hunk 3]
```diff
@@ -1489,8 +1494,10 @@ def _try_compile(
         for stmt in statements:
             comp, capabilities = self._compile_dispatch_ql(ctx, stmt)
 
+            drop_db = isinstance(comp, dbstate.DDLQuery) and comp.drop_db
```
[the remark this answers is not in the document]
Comment (1st1) [reply edb/server/compiler/compiler.py:1497]: good idea

**Review by elprans — APPROVED** [review]

**Pull request merged**